mailpoet/lib/Doctrine/WPDB/ConvertParameters.php

```diff
@@ -0,0 +1,84 @@
+<?php declare (strict_types = 1);
+
+namespace MailPoet\Doctrine\WPDB;
+
+use MailPoet\Doctrine\WPDB\Exceptions\MissingParameterException;
+use MailPoetVendor\Doctrine\DBAL\ParameterType;
+use MailPoetVendor\Doctrine\DBAL\SQL\Parser\Visitor;
+
+class ConvertParameters implements Visitor {
+  private const PARAM_TYPE_MAP = [
+    ParameterType::STRING => '%s',
+    ParameterType::INTEGER => '%d',
+    ParameterType::ASCII => '%s',
+    ParameterType::BINARY => '%s',
+    ParameterType::BOOLEAN => '%d',
+    ParameterType::NULL => '%s',
+    ParameterType::LARGE_OBJECT => '%s',
+  ];
+
+  /** @var list<string> */
+  private array $buffer = [];
+
+  /** @var array<array-key, array{0: string, 1: mixed, 2: int}> */
+  private array $params;
+
+  private array $values = [];
+
+  private int $cursor = 1;
+
+  public function __construct(
+    array $params
+  ) {
+    $this->params = $params;
+  }
+
+  public function acceptPositionalParameter(string $sql): void {
+    $position = $this->cursor++;
+    $this->acceptParameter($position);
+  }
+
+  public function acceptNamedParameter(string $sql): void {
+    $this->acceptParameter(trim($sql, ':'));
+  }
+
+  public function acceptOther(string $sql): void {
+    $this->buffer[] = $sql;
+  }
+
+  public function getSQL(): string {
+    return implode('', $this->buffer);
+  }
+
+  public function getValues(): array {
+    return $this->values;
+  }
+
+  /** @param array-key $key */
+  private function acceptParameter($key): void {
+    if (!array_key_exists($key, $this->params)) {
+      throw new MissingParameterException(sprintf("Parameter '%s' was defined in the query, but not provided.", $key));
+    }
+    [, $value, $type] = $this->params[$key];
+
+    // WPDB doesn't support NULL values. We need to handle them explicitly.
+    if ($value === null) {
+      $this->buffer[] = 'NULL';
+      return;
+    }
```

Comment:
This replicates the current behavior, but I see in the test it generates queries with conditions `WHERE column = null`, which is always false. I wonder how many such cases we have. I guess we can try to log this in our tests.

Reply:
@costasovo Did you manage to figure out how many of such cases we have in our codebase? That said, I think we'll need to keep this behavior to satisfy Doctrine's behavior. What I mean is that if the ORM can generate such queries, we should probably support them on the DBAL level as well.

Reply:
@JanJakes I ran the tests and logged all cases of null arguments that appeared after WHERE or JOIN. In the end, there weren't many places (actually 5), but they were called repeatedly, so the log looked quite big. 

Here is the code I used to check which cases to log to skip null in values:
```php
$shouldLog = false;
foreach ($this->buffer as $bufferPart) {
  if ((stripos($bufferPart, 'WHERE') !== false) || (stripos($bufferPart, 'JOIN') !== false)) {
    $shouldLog = true;
    break;
  }
}
```

I added two commits to address a few cases. Most of the logged occurrences came from queries created by high-level "findBy" methods, but Doctrine seems to be smart enough to cover those.
I found that, for example:
```php
$task = $this->scheduledTasksRepository->findOneBy(
      [
        'status' => [ScheduledTaskEntity::STATUS_SCHEDULED, null],
      ]
    );
```
is converted to
 ```sql
 (status IN ('scheduled', null) OR status IS NULL)
 ```
 
 I went through rest of the cases and added two commits, but they were more like nice to have.
 
 So I think we are fine.